datajunction-server/datajunction_server/models/access.py

```diff
@@ -0,0 +1,436 @@
+                if show_denials
+                else ""
+            )
+            raise DJException(
+                http_status_code=HTTPStatus.FORBIDDEN,
+                errors=[
+                    DJError(
+                        code=ErrorCode.UNAUTHORIZED_ACCESS,
+                        message=message,
+                    ),
+                ],
+            )
+
+    def validate(self) -> Set[ResourceRequest]:
+        """
+        Checks with ACS and stores any returned invalid requests
+        """
+        self.validation_request_count += 1
+
+        access_control = AccessControl(
+            user=self.user.username,
+            state=self.state,
+            direct_requests=deepcopy(self.direct_requests),
+            indirect_requests=deepcopy(self.indirect_requests),
+            validation_request_count=self.validation_request_count,
+        )
+        self.validate_access(access_control)  # type: ignore
+        self.validation_results = access_control.requests
+
+        if any((result.approved is None for result in self.validation_results)):
+            raise DJException(
+                http_status_code=HTTPStatus.FORBIDDEN,
+                errors=[
+                    DJError(
+                        code=ErrorCode.INCOMPLETE_AUTHORIZATION,
+                        message="Injected `validate_access` must approve or deny all requests.",
+                    ),
+                ],
+            )
+
+        return self.validation_results
+
+    def validate_and_raise(self):
+        """
+        Validates with ACS and raises if any resources were denied
+        """
+        self.validate()
+        self.raise_if_invalid_requests()
+
+
+def validate_access_nodes(
+    validate_access: "ValidateAccessFn",  # pylint: disable=W0621
+    verb: ResourceRequestVerb,
+    user: Optional[User],
+    nodes: Iterable[Node | NodeRevision],
+    raise_: bool = False,
+) -> List[NodeRevision | Node]:
+    """
+    Validate the access of the user to a set of nodes
+    """
+    if user is None:
+        raise DJException(#pragma: no cover
+            http_status_code=HTTPStatus.FORBIDDEN,
+            errors=[
+                DJError(
+                    code=ErrorCode.INCOMPLETE_AUTHORIZATION,
+                    message="Failed to acquire user to validate access.",
+                ),
+            ],
+        )
+    access_control = AccessControlStore(
+        validate_access=validate_access,
+        user=user,
+    )
+
+    access_control.add_request_by_nodes(verb, nodes)
+
+    validation_results = access_control.validate()
+    if raise_:
+        access_control.raise_if_invalid_requests()#pragma: no cover
+
+    return [
+        node
+        for node in nodes
+        if node.id
+        in {
+            request.access_object.revision_id
+            for request in validation_results
+            if request.approved
+        }
+    ]
+
+
+def validate_access_namespaces(
```

Comment:
nit: I think this models directory should only contain the model classes since I can imagine these authorization related functions will grow in number over time. These functions would probably be better located somewhere under the internal directory, maybe in a new submodule called `authorization`?